File: cast-extractor/src/main/java/com/libreshockwave/tools/format/InstructionFormatter.java
package com.libreshockwave.tools.format;

import com.libreshockwave.chunks.ScriptChunk;
import com.libreshockwave.chunks.ScriptNamesChunk;
import com.libreshockwave.format.ScriptFormatUtils;
import com.libreshockwave.lingo.Opcode;

/**
 * Formats bytecode instructions for display.
 */
public final class InstructionFormatter {

    private InstructionFormatter() {}

    /**
     * Formats a single bytecode instruction for display.
     */
    public static String format(ScriptChunk.Handler.Instruction instr, ScriptChunk script, ScriptNamesChunk names) {
        StringBuilder sb = new StringBuilder();
        sb.append(String.format("[%04d] %-16s", instr.offset(), instr.opcode().getMnemonic()));

        if (instr.rawOpcode() >= 0x40) {
            sb.append(" ");
            String argDesc = formatArgument(instr, script, names);
            sb.append(argDesc);
        }

        return sb.toString();
    }

    /**
     * Formats the argument of an instruction with contextual information.
     */
    public static String formatArgument(ScriptChunk.Handler.Instruction instr, ScriptChunk script, ScriptNamesChunk names) {
        int arg = instr.argument();
        var opcode = instr.opcode();
        String opName = opcode.name();

        // PUSH_CONS - literal constant (string, int, float)
        if (opcode == Opcode.PUSH_CONS) {
            if (arg >= 0 && arg < script.literals().size()) {
                var lit = script.literals().get(arg);
                String typeStr = ScriptFormatUtils.getLiteralTypeNameShort(lit.type());
                String valueStr = ScriptFormatUtils.formatLiteralValue(lit.value(), 40);
                return arg + " <" + typeStr + "> " + valueStr;
            }
        }

        // PUSH_SYMB - symbol name
        if (opcode == Opcode.PUSH_SYMB) {
            if (names != null && arg >= 0 && arg < names.names().size()) {
                return arg + " #" + names.getName(arg);
            }
        }

        // GET/SET variables - resolve names
        if (opName.startsWith("GET_") || opName.startsWith("SET_")) {
            if (opcode == Opcode.GET_GLOBAL ||
                opcode == Opcode.SET_GLOBAL ||
                opcode == Opcode.GET_GLOBAL2 ||
                opcode == Opcode.SET_GLOBAL2 ||
                opcode == Opcode.GET_PROP ||
                opcode == Opcode.SET_PROP ||
                opcode == Opcode.GET_OBJ_PROP ||
                opcode == Opcode.SET_OBJ_PROP ||
                opcode == Opcode.GET_MOVIE_PROP ||
                opcode == Opcode.SET_MOVIE_PROP ||
                opcode == Opcode.GET_TOP_LEVEL_PROP ||
                opcode == Opcode.GET_CHAINED_PROP) {
                if (names != null && arg >= 0 && arg < names.names().size()) {
                    return arg + " (" + names.getName(arg) + ")";
                }
            }
        }

        // Call opcodes - show function/handler name
        if (opcode == Opcode.LOCAL_CALL ||
            opcode == Opcode.EXT_CALL ||
            opcode == Opcode.OBJ_CALL ||
            opcode == Opcode.OBJ_CALL_V4 ||
            opcode == Opcode.TELL_CALL) {
            if (names != null && arg >= 0 && arg < names.names().size()) {
                return arg + " [" + names.getName(arg) + "]";
            }
        }

        // PUT/GET - resolve name
        if (opcode == Opcode.PUT ||
            opcode == Opcode.GET ||
            opcode == Opcode.SET) {
            if (names != null && arg >= 0 && arg < names.names().size()) {
                return arg + " (" + names.getName(arg) + ")";
            }
        }

        // THE_BUILTIN - show builtin name
        if (opcode == Opcode.THE_BUILTIN) {
            if (names != null && arg >= 0 && arg < names.names().size()) {
                return arg + " the " + names.getName(arg);
            }
        }

        // NEW_OBJ - parent script name
        if (opcode == Opcode.NEW_OBJ) {
            if (names != null && arg >= 0 && arg < names.names().size()) {
                return arg + " new(" + names.getName(arg) + ")";
            }
        }

        // PUSH_VAR_REF - variable reference
        if (opcode == Opcode.PUSH_VAR_REF) {
            if (names != null && arg >= 0 && arg < names.names().size()) {
                return arg + " @" + names.getName(arg);
            }
        }

        // Jump offsets - show target
        if (opcode.isJump()) {
            return arg + " -> offset " + arg;
        }

        return String.valueOf(arg);
    }
}

File: cpp/src/editor/format/InstructionFormatter.cpp
#include "libreshockwave/editor/format/InstructionFormatter.hpp"

#include <cstddef>
#include <iomanip>
#include <sstream>

#include "libreshockwave/chunks/ScriptNamesChunk.hpp"
#include "libreshockwave/format/ScriptFormatUtils.hpp"
#include "libreshockwave/lingo/Opcode.hpp"

namespace libreshockwave::editor::format {
namespace {

bool hasName(const chunks::ScriptNamesChunk* names, int index) {
    return names != nullptr && index >= 0 && index < static_cast<int>(names->names().size());
}

bool isNamedVariableOpcode(lingo::Opcode opcode) {
    switch (opcode) {
        case lingo::Opcode::GET_GLOBAL:
        case lingo::Opcode::SET_GLOBAL:
        case lingo::Opcode::GET_GLOBAL2:
        case lingo::Opcode::SET_GLOBAL2:
        case lingo::Opcode::GET_PROP:
        case lingo::Opcode::SET_PROP:
        case lingo::Opcode::GET_OBJ_PROP:
        case lingo::Opcode::SET_OBJ_PROP:
        case lingo::Opcode::GET_MOVIE_PROP:
        case lingo::Opcode::SET_MOVIE_PROP:
        case lingo::Opcode::GET_TOP_LEVEL_PROP:
        case lingo::Opcode::GET_CHAINED_PROP:
            return true;
        default:
            return false;
    }
}

bool isNamedCallOpcode(lingo::Opcode opcode) {
    switch (opcode) {
        case lingo::Opcode::LOCAL_CALL:
        case lingo::Opcode::EXT_CALL:
        case lingo::Opcode::OBJ_CALL:
        case lingo::Opcode::OBJ_CALL_V4:
        case lingo::Opcode::TELL_CALL:
            return true;
        default:
            return false;
    }
}

} // namespace

std::string InstructionFormatter::format(const chunks::ScriptChunk::Instruction& instruction,
                                         const chunks::ScriptChunk& script,
                                         const chunks::ScriptNamesChunk* names) {
    std::ostringstream out;
    out << "[" << std::setfill('0') << std::setw(4) << instruction.offset << "] "
        << std::setfill(' ') << std::left << std::setw(16) << lingo::mnemonic(instruction.opcode);
    if (instruction.rawOpcode >= 0x40) {
        out << " " << formatArgument(instruction, script, names);
    }
    return out.str();
}

std::string InstructionFormatter::formatArgument(const chunks::ScriptChunk::Instruction& instruction,
                                                 const chunks::ScriptChunk& script,
                                                 const chunks::ScriptNamesChunk* names) {
    const int argument = instruction.argument;
    const auto opcode = instruction.opcode;

    if (opcode == lingo::Opcode::PUSH_CONS) {
        const auto& literals = script.literals();
        if (argument >= 0 && argument < static_cast<int>(literals.size())) {
            const auto& literal = literals[static_cast<std::size_t>(argument)];
            return std::to_string(argument) + " <" +
                   ::libreshockwave::format::getLiteralTypeNameShort(literal.type) + "> " +
                   ::libreshockwave::format::formatLiteralValue(literal.value, 40);
        }
    }

    if (opcode == lingo::Opcode::PUSH_SYMB && hasName(names, argument)) {
        return std::to_string(argument) + " #" + names->getName(argument);
    }

    if (isNamedVariableOpcode(opcode) && hasName(names, argument)) {
        return std::to_string(argument) + " (" + names->getName(argument) + ")";
    }

    if (isNamedCallOpcode(opcode) && hasName(names, argument)) {
        return std::to_string(argument) + " [" + names->getName(argument) + "]";
    }

    if ((opcode == lingo::Opcode::PUT || opcode == lingo::Opcode::GET || opcode == lingo::Opcode::SET) &&
        hasName(names, argument)) {
        return std::to_string(argument) + " (" + names->getName(argument) + ")";
    }

    if (opcode == lingo::Opcode::THE_BUILTIN && hasName(names, argument)) {
        return std::to_string(argument) + " the " + names->getName(argument);
    }

    if (opcode == lingo::Opcode::NEW_OBJ && hasName(names, argument)) {
        return std::to_string(argument) + " new(" + names->getName(argument) + ")";
    }

    if (opcode == lingo::Opcode::PUSH_VAR_REF && hasName(names, argument)) {
        return std::to_string(argument) + " @" + names->getName(argument);
    }

    if (lingo::isJump(opcode)) {
        return std::to_string(argument) + " -> offset " + std::to_string(argument);
    }

    return std::to_string(argument);
}

} // namespace libreshockwave::editor::format
